arrow-fx-stm/src/main/kotlin/arrow/fx/stm/internal/Impl.kt

```diff
@@ -0,0 +1,214 @@
+   */
+  override fun <A> TVar<A>.read(): A =
+    when (val r = readVar(this as TVar<Any?>)) {
+      Entry.NOT_PRESENT -> readI().also { accessMap[this] = Entry(it, Entry.NO_CHANGE) }
+      else -> r as A
+    }
+
+  /**
+   * Add a write to the write set.
+   *
+   * If we have not seen this variable before we add a read which stores it in the read set as well.
+   */
+  override fun <A> TVar<A>.write(a: A): Unit =
+    accessMap[this as TVar<Any?>]?.update(a) ?: readI().let { accessMap[this] = Entry(it, a) }
+
+  internal fun validate(): Boolean =
+    accessMap.all { (tv, entry) -> tv.value === entry.initialVal }
+
+  internal fun validateAndCommit(): Boolean {
+    if (accessMap.isEmpty()) return true
+
+    val locked = mutableListOf<Map.Entry<TVar<Any?>, Entry>>()
+    val reads = mutableListOf<Map.Entry<TVar<Any?>, Entry>>()
+
+    /**
+     * Why do we not lock reads?
+     * To answer this question we need to ask under what conditions a transaction may commit:
+     * - A transaction can commit if all values read contain the same value when committing
+     *
+     * This means that when we hold all write locks we just need to verify that all our reads are consistent, any change after
+     *  that has no effect on this transaction because our write will 100% persist consistently (we hold all locks) and
+     *  any other transaction depending on a variable we are about to write to has to wait for us and then verify again
+     */
+    accessMap.forEach { tvToEntry ->
+      val (tv, entry) = tvToEntry
+      if (entry.isWrite()) {
+        if (tv.lock_cond(this, entry.initialVal)) {
+          locked.add(tvToEntry)
+        } else {
+          locked.forEach { it.key.release(this, it.value.initialVal) }
+          return@validateAndCommit false
+        }
+      } else {
+        if (tv.value !== entry.initialVal) {
+          locked.forEach { it.key.release(this, it.value.initialVal) }
+          return@validateAndCommit false
+        } else {
+          reads.add(tvToEntry)
+        }
+      }
+    }
+
+    if (reads.any { (tv, entry) -> tv.value !== entry.initialVal }) {
+      locked.forEach { it.key.release(this, it.value.initialVal) }
+      return false
+    }
+
+    locked.forEach { it.key.release(this, it.value.newVal) }
+    // TODO Evaluate if this needs to be separate or if it is cheap enough to do above.
+    //  Basically any work done before all locks are released needs to be cheap and this avoids a bit of work.
+    locked.forEach { it.key.notify() }
+    return true
+  }
+
+  private fun mergeReads(other: STMFrame): Unit {
+    accessMap.putAll(other.accessMap.filter { (_, e) -> e.isWrite().not() })
+  }
+
+  private fun merge(other: STMFrame): Unit {
+    accessMap.putAll(other.accessMap)
+  }
+}
+
+/**
+ * In some special cases it is possible to detect if a STM transaction blocks indefinitely so we can
+ *  abort here.
+ */
+object BlockedIndefinitely : Throwable("Transaction blocked indefinitely")
```

Comment:
Should these override fillInStackTrace from the Throwable constructor or is the stack trace meaningful here?

Reply:
For this one it does not matter I think. This is a terminal exception that'll usually kill the program as there is no sensible way to recover.

I might need to do that for the internal retry exception though...

Reply:
Which is to say, `BlockedIndefinitely` is not internal and control flow only and thus needs the stacktrace.

I changed the other internal only exception to override `fillInStackTrace` to a noop since it is only used for control flow anyway.

Also I am using exceptions instead of suspension because I only need aborts and never resume, so it is likely more efficient which is really important here...